File: smart-toast/src/main/java/com/coder/zzq/smartshow/toast/classic/ClassicToast.java
package com.coder.zzq.smartshow.toast.classic;

import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.annotation.IntDef;
import androidx.core.graphics.drawable.DrawableCompat;

import com.coder.zzq.smart_show.annotations.CustomToast;
import com.coder.zzq.smart_show.annotations.ToastConfig;
import com.coder.zzq.smart_show.annotations.ToastView;
import com.coder.zzq.smartshow.toast.Constants;
import com.coder.zzq.smartshow.toast.R;
import com.coder.zzq.smartshow.toast.factory.BaseToastConfig;
import com.coder.zzq.toolkit.Utils;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@CustomToast(alias = Constants.TOAST_TYPE_CLASSIC)
public class ClassicToast {

    @ToastConfig(withPrefix_m = true)
    public static class Config extends BaseToastConfig {

        public static final int ICON_POSITION_LEFT = 0;
        public static final int ICON_POSITION_RIGHT = 1;

        @Retention(RetentionPolicy.SOURCE)
        @IntDef({ICON_POSITION_LEFT, ICON_POSITION_RIGHT})
        @interface IconPosition {
        }

        @DrawableRes
        public int mBackgroundResource = Constants.DEFAULT_VALUE;
        @ColorInt
        public int mBackgroundColor = Constants.DEFAULT_VALUE;
        public float mMsgSize = 14;
        @ColorInt
        public int mMsgColor = Color.WHITE;
        public boolean mMsgBold = false;
        @DrawableRes
        public int mIcon = Constants.DEFAULT_VALUE;
        @IconPosition
        public int mIconPosition = ICON_POSITION_LEFT;
        public int mIconSizeDp = Constants.DEFAULT_VALUE;
        public int mIconPaddingDp = Constants.DEFAULT_VALUE;
    }

    @ToastView
    public static View provideToastView(View rootView, LayoutInflater inflater, Config config) {
        if (rootView == null) {
            rootView = inflater.inflate(R.layout.smart_show_classic_toast, null);
        }

        int bgResource = config.mBackgroundResource != Constants.DEFAULT_VALUE ?
                config.mBackgroundResource
                : R.drawable.smart_show_classic_toast_bg;

        rootView.setBackgroundResource(bgResource);

        if (config.mBackgroundColor != Constants.DEFAULT_VALUE){
            Drawable bg = rootView.getBackground().mutate();
            bg.mutate();
            DrawableCompat.setTint(bg, config.mBackgroundColor);
        }

        TextView msgView = rootView.findViewById(R.id.smart_toast_message);
        msgView.setText(config.mMsg);
        msgView.setTextColor(config.mMsgColor);
        msgView.setTextSize(config.mMsgSize);
        msgView.getPaint().setFakeBoldText(config.mMsgBold);


        Drawable icon = config.mIcon != Constants.DEFAULT_VALUE ? Utils.getDrawableFromRes(config.mIcon) : null;
        if (icon != null) {
            int iconSize = config.mIconSizeDp != Constants.DEFAULT_VALUE ?
                    config.mIconSizeDp
                    : (icon.getIntrinsicWidth() == -1 ? Math.round(config.mMsgSize) : Math.max(icon.getIntrinsicWidth(), icon.getIntrinsicHeight()));
            icon.setBounds(0, 0, iconSize, iconSize);
        }

        msgView.setCompoundDrawables(
                config.mIconPosition == Config.ICON_POSITION_LEFT ? icon : null,
                null,
                config.mIconPosition == Config.ICON_POSITION_RIGHT ? icon : null,
                null
        );

        msgView.setCompoundDrawablePadding(config.mIconPaddingDp != Constants.DEFAULT_VALUE ? config.mIconPaddingDp : Constants.DEFAULT_ICON_PADDING);
        return rootView;
    }
}

File: smart-toast/src/main/java/com/coder/vincent/smart_toast/alias/classic/ClassicToast.kt
package com.coder.vincent.smart_toast.alias.classic

import android.annotation.SuppressLint
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.ColorInt
import com.coder.vincent.series.annotations.CustomizedConfig
import com.coder.vincent.series.annotations.DataItem
import com.coder.vincent.series.annotations.ResourceType
import com.coder.vincent.series.annotations.smart_toast.CustomizedToast
import com.coder.vincent.series.common_lib.bean.IconPosition
import com.coder.vincent.series.common_lib.dpToPx
import com.coder.vincent.smart_toast.R
import com.coder.vincent.smart_toast.TOAST_ALIAS_CLASSIC
import com.coder.vincent.smart_toast.alias.ToastDefinition
import com.coder.vincent.smart_toast.databinding.SmartShowClassicToastBinding
import com.coder.vincent.smart_toast.factory.ToastConfig

@CustomizedToast(alias = TOAST_ALIAS_CLASSIC)
internal class ClassicToast : ToastDefinition<ClassicToast.Config> {

    @CustomizedConfig
    class Config : ToastConfig() {
        @ColorInt
        @DataItem(supportedResource = ResourceType.COLOR)
        var backgroundColor: Int = Color.parseColor("#cc000000")

        @DataItem
        var iconPosition = IconPosition.LEFT
    }


    @SuppressLint("InflateParams")
    override fun toastView(inflater: LayoutInflater): View =
        inflater.inflate(R.layout.smart_show_classic_toast, null)

    override fun applyConfig(toastView: View, config: Config) {
        SmartShowClassicToastBinding.bind(toastView).apply {
            val background = config.backgroundDrawable.mutate()
            if (background is GradientDrawable) {
                background.setColor(config.backgroundColor)
            }
            root.background = background
            smartToastMessage.text = config.message
            config.messageStyle.applyToView(smartToastMessage)
            val icon = config.iconDrawable?.apply {
                val iconSize = config.iconSize ?: (config.messageStyle.size + 3)
                setBounds(0, 0, iconSize.dpToPx(), iconSize.dpToPx())
            }
            val iconLeft = if (config.iconPosition == IconPosition.LEFT) icon else null
            val iconRight = if (config.iconPosition == IconPosition.RIGHT) icon else null
            smartToastMessage.setCompoundDrawables(iconLeft, null, iconRight, null)
            smartToastMessage.compoundDrawablePadding = config.marginBetweenIconAndMsg.dpToPx()
        }
    }
}